firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapDecoder.kt

```diff
@@ -0,0 +1,244 @@
+    }
+    protected var elementIndex: Int = START_INDEX
+    protected var isCurrentDecodeElementNotNull: Boolean = true
+    protected lateinit var decodedElementDataType: SerialKind
+
+    /** A list of values that need to be decoded as fields of the [Serializable] object. */
+    abstract val decodeValueList: List<Any?>
+
+    /** Decodes an enum field by returning its index from the enum object's descriptor. */
+    final override fun decodeEnum(enumDescriptor: SerialDescriptor): Int {
+        val decodedEnumFieldName = decodeValueList.elementAt(elementIndex - 1)
+        // TODO: Add a EnumNamingProperties parameter, and convert decodedEnumFieldName based on it
+        // i.e. case insensitive, snake_case match camelCase, etc
+        val enumFieldNames = enumDescriptor.elementNames.toList()
+        return enumFieldNames.indexOf(decodedEnumFieldName)
+    }
+
+    /**
+     * Returns true if the current element being decoded is not null.
+     *
+     * In case this method returns false, [decodeNull] method will be used instead of [decodeValue]
+     * to decode this null value element.
+     */
+    final override fun decodeNotNullMark(): Boolean {
+        val result = isCurrentDecodeElementNotNull
+        isCurrentDecodeElementNotNull = true
+        return result
+    }
+
+    /**
+     * Returns a not null primitive value that is going to be assigned as a field of the decoded
+     * [Serializable] object.
+     *
+     * Note: Firestore saves Int values as Long, and Float values as Double from the server side;
+     * therefore, during decoding process, depending on the field's [SerialKind], cast might be
+     * required to convert Long back to Int, and to convert Double back to Float.
+     */
+    final override fun decodeValue(): Any {
+        val element: Any = validateDecodedElementNotNull()
+        return when (decodedElementDataType) {
+            is PrimitiveKind.INT -> return (element as Long).toInt() // Firestore saves Int as Long
+            is PrimitiveKind.FLOAT ->
+                return (element as Double).toFloat() // Firestore saves Float as Double
+            else -> element
+        }
+    }
+
+    /**
+     * Returns a not null value from [decodeValueList]. This is the value that is going to be
+     * decoded as the field value of the [Serializable] object.
+     *
+     * Note: An [IllegalArgumentException] will be thrown if the [elementIndex] is pointing to a
+     * null value from [decodeValueList].
+     */
+    private fun validateDecodedElementNotNull(): Any =
+        decodeValueList.elementAt(elementIndex - 1)
+            ?: throw IllegalArgumentException(
+                "Got a null value while trying to decode a not null field."
+            )
+
+    final override val serializersModule: SerializersModule = EmptySerializersModule
+
+    final override fun beginStructure(descriptor: SerialDescriptor): CompositeDecoder {
+        val innerCompositeObject = getCompositeObject(elementIndex)
```

Comment:
It's very possible that the object at `elementIndex` is not composite. We, the ones with an object at `elementIndex` **are** composite for sure. Inner object is a better name.

Reply:
Actually, when this `beginStructure()` method is triggered, it is guaranteed from the `descriptor`'s meta data that the element need to be decoded is a composite element.